### DISCUSSION TITLE
Extract Base64 private/public Key Material from generated PGP KeyRing

### DISCUSSION THREAD

**Discussion by ghost** [pgpainless/pgpainless#279] [Q&A] [answered]
Hi @vanitasvitae, thank you very much for this library. What is the best way to read the keypair in Base64? This is what I've been trying:

```
 val pgp = PGPainless.buildKeyRing()
    .addUserId("someid")
            .setPrimaryKey(
                KeySpec.getBuilder(
                    ECDSA.fromCurve(EllipticCurve._SECP256K1),
                    KeyFlag.SIGN_DATA, KeyFlag.CERTIFY_OTHER
                )
            ).addSubkey(
                KeySpec.getBuilder(
                    ECDSA.fromCurve(EllipticCurve._SECP256K1),
                    KeyFlag.SIGN_DATA, KeyFlag.CERTIFY_OTHER
                )
            ).setPassphrase(Passphrase.emptyPassphrase())
            .build()
```

And after generating the keyring:

```
val sk = pgp.secretKeys.next()
        val pr = Hex.encode(sk.encoded).toString()
        val pb = Hex.encode(sk.publicKey.encoded).toString()
```

The output is longer than 32 bytes so I've tryied to guess where the keys could be. 
And looks like they are last 32 bytes and 64 bytes respectively for private and public keys. Is that right? Could you give advice me for a better way to do it? Thanks

**Comment by vanitasvitae** [comment, marked as answer]
I'm not totally sure to be honest.

I would guess it would be lie this:
```
PGPSecretKeyRing secretKeys = ...; // your generation code
// Iterate over all secret keys - there is more than one
for (PGPSecretKey secretKey : secretKeys) {
    PGPPublicKey publicKey = secretKey.getPublicKey();
    PGPPrivateKey privateKey = UnlockSecretKey.unlockSecretKey(secretKey, Passphrase.emptyPassphrase()); // Or whatever passphrase was used

    byte[] pubKeyBytes = pubKey.getPublicKeyPacket().getKey().getEncoded();
    byte[] secKeyBytes = privateKey.getPrivateKeyDataPacket().getEncoded();

    // You need to do the Base64 encoding yourselves, e.g. like this:
    String b64 = Base64.encodeToString(pubKeyBytes);
}
```

However, I'm not sure.

What is your usecase, that you need access to the raw key material?

Just to be sure, I guess you are aware of `PGPainless.asciiArmor(pgpSecretKeyRing)`, and this is not what you want?

**Reply by ghost** [reply]
It works, thanks. There is still a prefix on both keys wich I could not get the meaning. I'm studying the algorithm for an Android app and trying to understand if the ec curve is the same used to generate ethereum address. In the final app I would like to use this library to give an additional option to encrypt a google tink keyset and/or a symmectric key for sharing encrypted files.

**Reply by vanitasvitae** [reply]
If you just need some code to generate and use EC keys, why don't you simply use Bouncy Castle directly? PGPainless being an OpenPGP library adds a lot of unneeded overhead for you.

**Reply by ghost** [reply]
PGP support is still a nice feature, so that the user should be able to import existing keyring. Anyway I've tryied testing with BC but could not find the secp256k1 generator, maybe I've not imported all the necessary libraries or just got lost in the numerous amount of functions available

**Reply by vanitasvitae** [reply]
For the record, this would be done something like this:

```
KeyPairGenerator keyGenerator = KeyPairGenerator.getInstance("ECDSA", "BC");
keyGenerator.initialize(new ECNamedCurveGenParameterSpec("secp256k1"));
KeyPair keyPair = keyGenerator.generateKeyPair();
```
